libraries/ms-common/src/stm32f0xx/pwm_input.c

```diff
@@ -0,0 +1,201 @@
+    return;
+  }
+
+  TIM_ClearITPendingBit(tim_location, TIM_IT_CC1);
+
+  uint32_t IC2Value_1 = 0;
+  uint32_t IC2Value_2 = 0;
+
+  uint32_t period = 0;
+  uint32_t dc = 0;
+
+  if (s_port[timer].channel == TIM_Channel_2) {
+    IC2Value_1 = TIM_GetCapture2(tim_location);
+    IC2Value_2 = TIM_GetCapture1(tim_location);
+  } else {
+    IC2Value_2 = TIM_GetCapture2(tim_location);
+    IC2Value_1 = TIM_GetCapture1(tim_location);
+  }
+
+  if (IC2Value_1 != 0) {
+    dc = (IC2Value_2 * 1000) / IC2Value_1;
+    period = IC2Value_1;
+  } else {
+    dc = 0;
+    period = 0;
+  }
+
+  if (dc > MAX_DC_VALUE) {
+    dc = MAX_DC_VALUE;
+  }
+
+  s_port[timer].period = period;
+  s_port[timer].dc = dc;
+
+  if (s_port[timer].storage->callback != NULL) {
+    s_port[timer].storage->callback(dc, period, s_port[timer].storage->context);
+  }
+}
+
+StatusCode pwm_input_init(PwmTimer timer, PwmInputSettings *settings, PwmInputStorage *storage) {
+  s_port[timer].rcc_cmd(s_port[timer].periph, ENABLE);
+
+  memset(storage, 0, sizeof(*storage));
+  s_port[timer].storage = storage;
+  s_port[timer].storage->callback = settings->callback;
+  s_port[timer].storage->context = settings->context;
+  s_port[timer].channel = settings->channel;
+
+  uint16_t trigger_source = 0;
+
+  if (settings->channel == PWM_CHANNEL_1) {
+    s_port[timer].channel = TIM_Channel_1;
+    trigger_source = TIM_TS_TI1FP1;
+  } else {
+    s_port[timer].channel = TIM_Channel_2;
+    trigger_source = TIM_TS_TI2FP2;
+  }
+
+  stm32f0xx_interrupt_nvic_enable(s_port[timer].irq, 0);
+
+  TIM_TypeDef *tim_location = s_port[timer].base;
+
+  RCC_ClocksTypeDef clocks;
+  RCC_GetClocksFreq(&clocks);
+
+  TIM_TimeBaseInitTypeDef tim_init = {
+    .TIM_Prescaler = (clocks.PCLK_Frequency / 1000000) - 1,
+    .TIM_CounterMode = TIM_CounterMode_Up,
+    .TIM_Period = 0xFFFFFFFF,
+    .TIM_ClockDivision = TIM_CKD_DIV1,
+    .TIM_RepetitionCounter = 0,
+  };
+
+  TIM_TimeBaseInit(tim_location, &tim_init);
+
+  TIM_ICInitTypeDef tim_icinit = {
+    .TIM_Channel = s_port[timer].channel,
+    .TIM_ICPolarity = TIM_ICPolarity_Rising,
+    .TIM_ICSelection = TIM_ICSelection_DirectTI,
+    .TIM_ICPrescaler = TIM_ICPSC_DIV1,
+    .TIM_ICFilter = 0x0,
+  };
+
+  TIM_PWMIConfig(tim_location, &tim_icinit);
+
+  TIM_SelectInputTrigger(tim_location, trigger_source);
+  TIM_SelectSlaveMode(tim_location, TIM_SlaveMode_Reset);
+  TIM_SelectMasterSlaveMode(tim_location, TIM_MasterSlaveMode_Enable);
+  TIM_SelectOutputTrigger(tim_location, TIM_TRGOSource_Reset);
+  TIM_Cmd(tim_location, ENABLE);
+  TIM_ITConfig(tim_location, TIM_IT_CC1, ENABLE);
+
+  return STATUS_CODE_OK;
+}
+
+uint32_t pwm_input_get_period(PwmTimer timer) {
+  CRITICAL_SECTION_AUTOEND;
+  uint32_t result = s_port[timer].period;
+  s_port[timer].period = 0;
+  return result;
+}
+
+uint32_t pwm_input_get_dc(PwmTimer timer) {
+  CRITICAL_SECTION_AUTOEND;
+  uint32_t result = s_port[timer].dc;
+
+  if (result == 0) {
```

Comment:
I wouldn't expect a function called `pwm_get_input` to run a callback if the DC is 0; that seems like a weird side-effect to have. IMO it would be more consistent to have this function just return the DC and have another function that always fires the callback (even if the DC is 0) that way the behavior is more reasonable/transparent.

Reply:
Ok, that's fair. It's violating the principle of least astonishment 😅. I'll just remove callbacks altogether though, and it's should be an issue